File: day4/day4.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

#include "../api.h"

bool is_anagram(char* first, char* second) {
	int firstLetters[26] = {0};
	int secondLetters[26] = {0};

	for (int i = 0; i < strlen(first); i++) {
		firstLetters[first[i] - 'a']++;
	}

	for (int i = 0; i < strlen(second); i++) {
		secondLetters[second[i] - 'a']++;
	}

	for (int i = 0; i < 26; i++) {
		if (firstLetters[i] != secondLetters[i]) {
			return false;
		}
	}
	return true;
}

bool is_valid(char* phrase, bool (*condition)(char*, char**, int)) {
	SplitResult* splits = split(phrase, ' ');

	char** words = malloc(sizeof(char*) * splits->count);

	for (int i = 0; i < splits->count; i++) {
		if (!condition(splits->strings[i], words, i)) {
			return false;
		}
		words[i] = splits->strings[i];
	}

	return true;
}

bool is_in(char* word, char** words, int wordCount) {
	for (int i = 0; i < wordCount; i++) {
		if (strcmp(word, words[i]) == 0) {
			return false;
		}
	}
	return true;
}

bool has_anagram(char* word, char** words, int wordCount) {
	for (int i = 0; i < wordCount; i++) {
		if (is_anagram(word, words[i])) {
			return false;
		}
	}
	return true;
}

int main(int argc, char** argv) {
	Input* input = load_input();

	int part1 = 0;
	int part2 = 0;

	for (int i = 0; i < input->count; i++) {
		if (is_valid(input->lines[i], is_in)) {
			part1++;
		}
		if (is_valid(input->lines[i], has_anagram)) {
			part2++;
		}
	}

	printf("%d\n", part1);
	printf("%d\n", part2);

	return 0;
}

File: day4/day4.java
import java.io.BufferedInputStream;
import java.util.*;

public class day4 {
    private static boolean isValid(String phrase, Validator validator) {
        String[] strings = phrase.split(" ");
        String[] words = new String[strings.length];
        int count = 0;
        for (int i = 0; i < strings.length; i++) {
            if (!validator.validate(strings[i], words)) {
                return false;
            }
            words[count++] = strings[i];
        }
        return true;
    }

    private static boolean isAnagram(String first, String second) {
        char[] firsts = first.toCharArray();
        if (second == null) {
            return false;
        }
        char[] seconds = second.toCharArray();
        Arrays.sort(firsts);
        Arrays.sort(seconds);
        return Arrays.equals(firsts, seconds);
    }

    public static void main(String[] args) {
        Scanner scan = new Scanner(new BufferedInputStream(System.in));

        Validator isIn = (String word, String[] words) -> {
            for (int i = 0; i < words.length; i++) {
                if (word.equals(words[i])) {
                    return false;
                }
            }
            return true;
        };

        Validator isAnagram = (String word, String[] words) -> {
            for (int i = 0; i < words.length; i++) {
                if (isAnagram(word, words[i])) {
                    return false;
                }
            }
            return true;
        };

        int validOne = 0;
        int validTwo = 0;
        while (scan.hasNext()) {
            String line = scan.nextLine();
            if (isValid(line, isIn)) {
                validOne += 1;
            }
            if (isValid(line, isAnagram)) {
                validTwo += 1;
            }
        }

        System.out.println(validOne);
        System.out.println(validTwo);
    }
}

interface Validator {
    boolean validate(String word, String[] words);
}